File: seira3/Anagrams.java
import java.io.*;
import java.util.*;

/* The objects of this class describe the state of the system 
	on a given time. */
class State {
	String Stack1;		// Contents of stack 1
	String Stack2;		// Contents of stack 2
	String Buffer;		// Contents of buffer

	/* Constructor */
	State(String newStack1, String newStack2, String newBuffer) {
		this.Stack1 = newStack1;
		this.Stack2 = newStack2;
		this.Buffer = newBuffer;
	}

	/* Returns the contents of stack 1 */
	String getStack1() {
		return this.Stack1;
	}
		
	/* Returns the contents of stack 2 */
	String getStack2() {
		return this.Stack2;
	}

	/* Returns the contents of the buffer */
	String getBuffer() {
		return this.Buffer;
	}

	/* Overriding the hashCode() and equals() methods
		so that the operations on the hashmap 'checked' work properly */
	@Override
	public int hashCode() {
		return Stack1.hashCode() ^ Stack2.hashCode() ^ Buffer.hashCode();
	}

	@Override
	public boolean equals(Object obj) {
		if (obj == null)
			return false;
		if (!(obj instanceof State))
			return false;
		
		State other = (State) obj;

		if (!this.Stack1.equals(other.Stack1))
			return false;
		if (!this.Stack2.equals(other.Stack2))
			return false;
		if (!this.Buffer.equals(other.Buffer))
			return false;

		return true;
	}
}

public class Anagrams {
	static HashMap<State, String> checked = new HashMap<State, String>();
	/* Keeps the states that we've already been through, so that
		we avoid making circles.
		Structure:  {[Stack1, Stack2, Buffer] -> [Moves]} */

	static ArrayList<String[]> tree = new ArrayList<String[]>();
	/* Keeps the states opened by the BFS algorithm.
		Structure: ArrayList of [Stack1, Stack2, Buffer, Moves] */

	static String target = new String();
	/* The second string given */

	static boolean found = false;

	/* Performs a 10-move from the given state and returns the 
		new state */
	public static State move10(String st1, String st2, String buff) {
		if ((st1.equals("")) || (!buff.equals(""))) {
			return null;
		} else {
			buff = Character.toString(st1.charAt(st1.length()-1));
			if (st1.length() == 1)
				st1 = "";
			else
				st1 = st1.substring(0, st1.length()-1);

			State nextState = new State(st1, st2, buff);

			if (!checked.containsKey(nextState))
				return nextState;
			else
				return null;
		}
	}

	/* Performs a 12-move from the given state and returns the 
		new state */
	public static State move12(String st1, String st2, String buff) {
		if (st1.equals(""))
			return null;
		else {
			st2 = st2.concat(Character.toString(st1.charAt(st1.length()-1)));

			if (st2.equals(target))
				found = true;

			if (st1.length() == 1)
				st1 = "";
			else
				st1 = st1.substring(0, st1.length()-1);

			State nextState = new State(st1, st2, buff);

			if (!checked.containsKey(nextState))
				return nextState;
			else
				return null;
		}
	}

	/* Performs a 01-move from the given state and returns the 
		new state */
	public static State move01(String st1, String st2, String buff) {
		if (buff.equals(""))
			return null;
		else {
			State nextState = new State(st1.concat(buff), st2, "");

			if (!checked.containsKey(nextState))
				return nextState;
			else
				return null;
		}			
	}

	/* Performs a 02-move from the given state and returns the 
		new state */
	public static State move02(String st1, String st2, String buff) {
		if (buff.equals(""))
			return null;
		else {
			st2 = st2.concat(buff);
			State nextState = new State(st1, st2, "");

			if (st2.equals(target))
				found = true;

			if (!checked.containsKey(nextState)) 
				return nextState;
			else
				return null;
		}
	}

	/* Performs a 20-move from the given state and returns the 
		new state */
	public static State move20(String st1, String st2, String buff) {
		if (st2.equals("") || !buff.equals("")) 
			return null;
		else {
			buff = Character.toString(st2.charAt(st2.length()-1));
			if (st2.length() == 1)
				st2 = "";
			else
				st2 = st2.substring(0, st2.length()-1);

			if (st2.equals(target))
				found = true;

			State nextState = new State(st1, st2, buff);

			if (!checked.containsKey(nextState))
				return nextState;
			else
				return null;
		}
	}

	/* Performs a 21-move from the given state and returns the 
		new state */
	public static State move21(String st1, String st2, String buff) {
		if (st2.equals(""))
			return null;
		else {
			st1 = st1.concat(Character.toString(st2.charAt(st2.length()-1)));
			if (st2.length() == 1)
				st2 = "";
			else
				st2 = st2.substring(0, st2.length()-1);
			
			if (st2.equals(target))
				found = true;

			State nextState = new State(st1, st2, buff);

			if (!checked.containsKey(nextState))
				return nextState;
			else
				return null;
		}
	}

	/* The main method */
	public static void main(String[] args) {
		// Check input
		if (args.length == 0) {		
			System.out.println("Error: Not enough input arguments\n");
			System.exit(1);
		}

		State root = new State(args[0], "", "");
		target = args[1];

		String[] initial = new String[4];
		initial[0] = args[0];
		initial[1] = "";
		initial[2] = "";
		initial[3] = "";

		// Add initial state to tree
		tree.add(initial);

		// Perform BFS looking for the target string
		String st1 = args[0];
		String st2 = "";
		String buff = "";
		String moves = "";

		while (!st2.equals(target)) {
			st1 = tree.get(0)[0];
			st2 = tree.get(0)[1];
			buff = tree.get(0)[2];
			moves = tree.get(0)[3];

			tree.remove(0);

			if (st2.equals(target)) { 	// Stop search - target reached
				System.out.println(moves);
				break;
			} else {		// Keep searching
				State move10 = move10(st1, st2, buff);
				if (move10 instanceof State) {
					String newMoves;
					if (moves.equals(""))
						newMoves = "10";
					else
						newMoves = moves.concat("-10");

					String[] newState = new String[4];
					newState[0] = move10.getStack1();
					newState[1] = move10.getStack2();
					newState[2] = move10.getBuffer();
					newState[3] = newMoves;
					tree.add(newState);
					checked.put(move10, newMoves);
				}

				State move12 = move12(st1, st2, buff);
				if (move12 instanceof State) {
					String newMoves;
					if (moves.equals(""))
						newMoves = "12";
					else
						newMoves = moves.concat("-12");

					if (found) {
						System.out.println(newMoves);
						break;
					}

					String[] newState = new String[4];
					newState[0] = move12.getStack1();
					newState[1] = move12.getStack2();
					newState[2] = move12.getBuffer();
					newState[3] = newMoves;
					tree.add(newState);
					checked.put(move12, newMoves);
				}

				State move01 = move01(st1, st2, buff);
				if (move01 instanceof State) {
					String newMoves;
					if (moves.equals(""))
						newMoves = "01";
					else
						newMoves = moves.concat("-01");
					String[] newState = new String[4];
					newState[0] = move01.getStack1();
					newState[1] = move01.getStack2();
					newState[2] = move01.getBuffer();
					newState[3] = newMoves;
					tree.add(newState);
					checked.put(move01, newMoves);
				}

				State move02 = move02(st1, st2, buff);
				if (move02 instanceof State) {
					String newMoves;
					if (moves.equals(""))
						newMoves = "02";
					else
						newMoves = moves.concat("-02");
				
					if (found) {
						System.out.println(newMoves);
						break;
					}

					String[] newState = new String[4];
					newState[0] = move02.getStack1();
					newState[1] = move02.getStack2();
					newState[2] = move02.getBuffer();
					newState[3] = newMoves;
					tree.add(newState);
					checked.put(move02, newMoves);
				}

				State move20 = move20(st1, st2, buff);
				if (move20 instanceof State) {
					String newMoves;
					if (moves.equals(""))
						newMoves = "20";
					else
						newMoves = moves.concat("-20");

					if (found) {
						System.out.println(newMoves);
						break;
					}

					String[] newState = new String[4];
					newState[0] = move20.getStack1();
					newState[1] = move20.getStack2();
					newState[2] = move20.getBuffer();
					newState[3] = newMoves;
					tree.add(newState);
					checked.put(move20, newMoves);
				}

				State move21 = move21(st1, st2, buff);
				if (move21 instanceof State) {
					String newMoves;
					if (moves.equals(""))
						newMoves = "21";
					else
						newMoves = moves.concat("-21");

					if (found) {
						System.out.println(newMoves);
						break;
					}

					String[] newState = new String[4];
					newState[0] = move21.getStack1();
					newState[1] = move21.getStack2();
					newState[2] = move21.getBuffer();
					newState[3] = newMoves;
					tree.add(newState);
					checked.put(move21, newMoves);
				}
			}
		}
	}
	
}

File: seira3/anagrams.pl
%	?- move10/4
%	
%	Returns the next 10-move from the given state and
%	updates the Checked associated list.
%
move10([_, Stack1, _]-_, Checked, Checked, B-B) :-
	dlist2list(Stack1, []).
move10([Stack2, [Stack1H|Stack1T]-Trail, []]-Moves, Checked, NewChecked, [[Stack2, Stack1T-Trail, [Stack1H]]-['10'|Moves]|B]-B) :-
	duplicate_term(Stack2, Stack2_Dup),
	duplicate_term(Stack1T-Trail, Stack1_Dup),
	dlist2list(Stack2_Dup, Stack2L),
	dlist2list(Stack1_Dup, Stack1L),
	not(get_assoc([Stack2L, Stack1L, [Stack1H]], Checked, _)),
	put_assoc([Stack2L, Stack1L, [Stack1H]], Checked, ['10'|Moves], NewChecked).
move10([_, _, _]-_, Checked, Checked, B-B).
	
%	?- move12/4
%	
%	Returns the next 12-move from the given state and
%	updates the Checked associated list.
%
move12([_, Stack1, _]-_, Checked, Checked, B-B) :-
	dlist2list(Stack1, []).
move12([Stack2, [Stack1H|Stack1T]-Trail, Buffer]-Moves, Checked, NewChecked, [[NewStack2, Stack1T-Trail, Buffer]-['12'|Moves]|B]-B) :-
	conc([Stack1H|A]-A, Stack2, NewStack2),
	duplicate_term(NewStack2, NewStack2_Dup),
	duplicate_term(Stack1T-Trail, Stack1_Dup),
	dlist2list(NewStack2_Dup, NewStack2L),
	dlist2list(Stack1_Dup, Stack1L),
	not(get_assoc([NewStack2L, Stack1L, Buffer], Checked, _)),
	put_assoc([NewStack2L, Stack1L, Buffer], Checked, ['12'|Moves], NewChecked).
move12([_,_,_]-_, Checked, Checked, B-B).

%	?- move01/4
%	
%	Returns the next 01-move from the given state and
%	updates the Checked associated list.
%
move01([_, _, []]-_, Checked, Checked, B-B).
move01([Stack2, Stack1-Trail, [Buffer]]-Moves, Checked, NewChecked, [[Stack2, [Buffer|Stack1]-Trail, []]-['01'|Moves]|B]-B) :-
	duplicate_term(Stack2, Stack2_Dup),
	duplicate_term(Stack1-Trail, Stack1_Dup),
	dlist2list(Stack2_Dup, Stack2L),
	dlist2list(Stack1_Dup, Stack1L),
	not(get_assoc([Stack2L, [Buffer|Stack1L], []], Checked, _)),
	put_assoc([Stack2L, [Buffer|Stack1L], []], Checked, ['01'|Moves], NewChecked).
move01([_,_,_]-_, Checked, Checked, B-B).

%	?- move02/4
%	
%	Returns the next 02-move from the given state and
%	updates the Checked associated list.
%
move02([_, _, []]-_, Checked, Checked, B-B).
move02([Stack2-Trail, Stack1, [Buffer]]-Moves, Checked, NewChecked, [[[Buffer|Stack2]-Trail, Stack1, []]-['02'|Moves]|B]-B) :-
	duplicate_term(Stack1, Stack1_Dup),
	duplicate_term(Stack2-Trail, Stack2_Dup),
	dlist2list(Stack1_Dup, Stack1L),
	dlist2list(Stack2_Dup, Stack2L),
	not(get_assoc([[Buffer|Stack2L], Stack1L, []], Checked, _)),
	put_assoc([[Buffer|Stack2L], Stack1L, []], Checked, ['02'|Moves], NewChecked).
move02([_,_,_]-_, Checked, Checked, B-B).

%	?- move21/4
%	
%	Returns the next 21-move from the given state and
%	updates the Checked associated list.
%
move21([Stack2, _, _]-_, Checked, Checked, B-B) :-
	dlist2list(Stack2, []).
move21([[Stack2H|Stack2T]-Trail, Stack1, Buffer]-Moves, Checked, NewChecked, [[Stack2T-Trail, NewStack1, Buffer]-['21'|Moves]|B]-B) :-
	conc([Stack2H|A]-A, Stack1, NewStack1),
	duplicate_term(NewStack1, NewStack1_Dup),
	duplicate_term(Stack2T-Trail, Stack2_Dup),
	dlist2list(NewStack1_Dup, NewStack1L),
	dlist2list(Stack2_Dup, Stack2L),
	not(get_assoc([Stack2L, NewStack1L, Buffer], Checked, _)),
	put_assoc([Stack2L, NewStack1L, Buffer], Checked, ['02'|Moves], NewChecked).
move21([_,_,_]-_, Checked, Checked, B-B).

%	?- move20/4
%	
%	Returns the next 20-move from the given state and
%	updates the Checked associated list.
%
move20([Stack2, _, _]-_, Checked, Checked, B-B) :-
	dlist2list(Stack2, []).
move20([[Stack2H|Stack2T]-Trail, Stack1, []]-Moves, Checked, NewChecked, [[Stack2T-Trail, Stack1, [Stack2H]]-['20'|Moves]|B]-B) :-
	duplicate_term(Stack1, Stack1_Dup),
	duplicate_term(Stack2T-Trail, Stack2_Dup),
	dlist2list(Stack1_Dup, Stack1L),
	dlist2list(Stack2_Dup, Stack2L),
	not(get_assoc([Stack2L, Stack1L, [Stack2H]], Checked, _)),
	put_assoc([Stack2L, Stack1L, [Stack2H]], Checked, ['20'|Moves], NewChecked).
move20([_, _, _]-_, Checked, Checked, B-B).

%	?- extend/4
%
% 	Returns a list of the next possible moves from the given state.
%
extend(Node, Checked, NewChecked, NewNodes) :-
	move10(Node, Checked, NewChecked1, Move10),
	move12(Node, NewChecked1, NewChecked2, Move12),
	move01(Node, NewChecked2, NewChecked3, Move01),
	move02(Node, NewChecked3, NewChecked4, Move02),
	move20(Node, NewChecked4, NewChecked5, Move20),
	move21(Node, NewChecked5, NewChecked, Move21), !,
	conc(Move10, Move12, M1),
	conc(M1, Move01, M2),
	conc(M2, Move02, M3),
	conc(M3, Move20, M4),
	conc(M4, Move21, NewNodes).
	

%	?- bfs/4
%
%	Performs a BFS algorithm on a given tree of states.

%	Use as bfs(Tree, Target, Checked, Moves), where:
%		Tree: a list representing the tree of states.
%			Its elements are of the form: [Stack2, Stack1, Buffer]-PathOfMovesToHere
%		Target: the string we are looking for	
%		Checked: an associated list of states that we have already been through (so we avoid making circles)
%		Moves: the (reversed) path of moves needed to get to the target string
%
bfs([[Stack2, _, _]-Moves|_]-_, Target, _, Moves) :-
	duplicate_term(Stack2, Stack2_Dup),
	dlist2list(Stack2_Dup, Stack2L),
	Stack2L \== [],
	Stack2 = Target.
bfs([[Stack2, Stack1, Buffer]-M|Rest]-Tail, Target, Checked, Moves) :-
	duplicate_term(Stack2, Stack2_Dup),
	duplicate_term(Target, Target_Dup),
	dlist2list(Stack2_Dup, Stack2List),
	dlist2list(Target_Dup, TargetList),
	Stack2List \== TargetList,
	extend([Stack2,Stack1,Buffer]-M, Checked, NewChecked, NewH),
	conc(Rest-Tail, NewH, NewList),
	bfs(NewList, Target, NewChecked, Moves).


%	?- getListRev/2
%
%	Use as getList(?Codes, ?List), where:
%		Codes: a list of codes representing characters
%		List: a (reversed) list of characters that correspond to the given codes
%
%	Uses the predicate getCharRev/2
%
getCharRev([X], [Character]) :-
	char_code(Character, X).
getCharRev([H|T], List) :-
	char_code(Character, H),
	getCharRev(T, Rest),
	append(Rest, [Character], List).

getListRev(Str, L-Tail) :-
	getCharRev(Str, List),
	append(List, Tail, L).

dlist2list(DiffList, List) :-
	convert2list(DiffList, List).

convert2list(L-[], L).

conc(A-B, B-C, A-C).

%	?- anagrams/3
%
%	anagrams(+Word, +Target, ?Moves)
%
anagrams(Word, Target, Moves) :-
	getListRev(Word, WordListDL),
	getListRev(Target, TargetListDL),
	list_to_assoc([], NewChecked),
	duplicate_term(WordListDL, WordListDL2),
	dlist2list(WordListDL2, WordList),
	put_assoc([[], WordList, []], NewChecked, [], Checked),
	bfs([[B-B, WordListDL, []]-[]|C]-C, TargetListDL, Checked, MovesRev),
	reverse(MovesRev, Moves).
